cql3/statements/create_table_statement.cc

```diff
@@ -205,18 +206,30 @@ std::unique_ptr<prepared_statement> create_table_statement::raw_statement::prepa
 
     auto stmt = ::make_shared<create_table_statement>(_cf_name, _properties.properties(), _if_not_exists, _static_columns, _properties.properties()->get_id());
 
-    std::optional<std::map<bytes, data_type>> defined_multi_cell_collections;
+    std::optional<std::map<bytes, data_type>> defined_multi_cell_columns;
     for (auto&& entry : _definitions) {
         ::shared_ptr<column_identifier> id = entry.first;
         cql3_type pt = entry.second->prepare(db, keyspace());
         if (pt.is_counter() && !service::get_local_storage_service().cluster_supports_counters()) {
             throw exceptions::invalid_request_exception("Counter support is not enabled");
         }
-        if (pt.is_collection() && pt.get_type()->is_multi_cell()) {
-            if (!defined_multi_cell_collections) {
-                defined_multi_cell_collections = std::map<bytes, data_type>{};
+        if (pt.get_type()->is_multi_cell()) {
+            // check for multi-cell types (non-frozen UDTs or collections) inside a non-frozen UDT
+            if (pt.get_type()->is_user_type()) {
+                auto type = static_cast<const user_type_impl*>(pt.get_type().get());
+                for (auto&& inner: type->all_types()) {
+                    if (inner->is_multi_cell()) {
+                        // a nested non-frozen UDT should have already been rejected when defining the type
```

Comment:
Why are non-frozen UDT rejected when defining the type but non-frozen collections are not? Shouldn't they be treated the same way?

Reply:
That's how Cassandra does it :(